source/pervasive/nonlinear_arith/div.rs

```diff
@@ -0,0 +1,1088 @@
+    {
+        lemma_basic_div(d);
+    }
+}
+
+/// numerical order is preserved when dividing two seperate integers by a common positive divisor
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_is_ordered(x: int, y: int, z: int)
+    requires 
+        x <= y,
+        0 < z,
+    ensures x / z <= y / z
+{
+    lemma_div_auto(z);
+    let f = |xy: int| xy <= 0 ==> (xy + y) / z <= y / z;
+
+    assert forall |i: int| #[trigger]is_le(i + 1, z) && f(i) implies f(i - z) by {
+        if (i - z <= 0) {
+            assert( f(i) );
+            assert( i <= 0 ==> (i + y) / z <= y / z );
+            if (i > 0) {
+                assert ( z > 0 );
+                assert( i <= z );
+                assert ( ((i + y) - z) / z <= y / z );
+            } else {
+                assert( (i + y) / z <= y / z );
+            }
+            assert( (i - z + y) / z <= y / z );
+        }
+    };
+
+    lemma_div_induction_auto(z, x - y, |xy: int| xy <= 0 ==> (xy + y) / z <= y / z);
+}
+
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_is_ordered_auto()
+    ensures forall |x: int, y: int, z: int| x <= y && 0 < z ==> #[trigger](x / z) <= #[trigger](y / z)
+{
+    assert forall |x: int, y: int, z: int| x <= y && 0 < z implies #[trigger](x / z) <= #[trigger](y / z) by
+    {
+        lemma_div_is_ordered(x, y, z);
+    }
+}
+
+/// dividing an integer by 2 or more results in a quotient that is smaller than the 
+/// original dividend
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_decreases(x: int, d: int)
+    requires 
+        0 < x,
+        1 < d,
+    ensures 
+        x / d  < x
+{
+    lemma_div_induction_auto(d, x, |u: int| 0 < u ==> u / d < u);
+}
+
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_decreases_auto()
+    ensures 
+        forall |x: int, d: int| 0 < x && 1 < d ==> #[trigger](x / d) < x,
+{
+    assert forall |x: int, d: int| 0 < x && 1 < d implies #[trigger](x / d) < x by
+    {
+        lemma_div_decreases(x, d);
+    };
+}
+
+/// dividing an integer by 1 or more results in a quotient that is less than or equal to 
+/// the original dividend
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_nonincreasing(x: int, d: int)
+    requires 
+        0 <= x,
+        0 < d,
+    ensures 
+        x / d  <= x
+{
+    lemma_div_induction_auto(d, x, |u: int| 0 <= u ==> u / d <= u);
+}
+
+proof fn lemma_div_nonincreasing_auto()
+    ensures forall |x: int, d: int| 0 <= x && 0 < d ==> #[trigger](x / d) <= x,
+{
+    assert forall |x: int, d: int| 0 <= x && 0 < d implies #[trigger](x / d) <= x by
+    {
+        lemma_div_nonincreasing(x, d);
+    }
+}
+
+/// a natural number x divided by a larger natural number gives a remainder equal to x
+// #[verifier::spinoff_prover]
+pub proof fn lemma_small_mod(x: nat, m: nat)
+    requires 
+        x < m,
+        0 < m
+    ensures 
+        x % m == x
+{
+    ModINL::lemma_small_mod(x, m);
+}
+
+/// The product of two positive integers is positive
+/// 
```

Comment:
I don't think this part of the `ensures` clause is really important and can be removed; that part of the ensures clause exists mostly just to allow the latter part of the ensures clause to work out to be reasonable; only the latter part is interesting, and should be included in this doc comment
```suggestion
```